src/logprod.jl

```diff
@@ -0,0 +1,78 @@
+"""
+    logprod(X::AbstractArray{T}; dims)
+
+Compute `sum(log.(X))` with a single `log` evaluation,
+provided `float(T) <: AbstractFloat`.
+
+This is faster than computing `sum(log, X)`, especially for large `X`.
+It works by representing the `j`th element of `x` as ``x_j = a_j  2^{b_j}`,
+allowing us to write
+```math
+\\sum_j \\log{x_j} = \\log(\\prod_j a_j) + \\log{2} \\sum_j b_j
+```
+"""
+logprod(x) = first(logabsprod(x))
+
+export logabsprod
+
+function logabsprod(x::AbstractArray{T}) where {T}
+    sig, ex = mapreduce(_logabsprod_op, x; init=frexp(one(T))) do xj
+        float_xj = float(xj)
+        frexp(float_xj)
+    end
+    sgn = signbit(sig) ? -one(T) : one(T)
+    return (log(abs(sig)) + IrrationalConstants.logtwo * T(ex), sgn)
+end
+
+@inline function _logabsprod_op((sig1, ex1), (sig2, ex2))
+    sig = sig1 * sig2
+    # sig = ifelse(sig2<0, sig2, sig1 * sig2)
+    ex = ex1 + ex2
+    # Significands are in the range [1,2), so multiplication will eventually overflow
+    if sig > floatmax(typeof(sig)) / 2
+        (new_sig, Δex) = frexp(sig)
+        ex += Δex
+        sig = new_sig
+    end
+    return sig, ex
+end
+
+"""
+    logprod(x)
+    logprod(f, x, ys...)
+
+For any iterator which produces `AbstractFloat` elements,
+this can use `logprod`'s fast reduction strategy.
+
+Signature with `f` is equivalent to `sum(log, map(f, x, ys...))`
+or `mapreduce(log∘f, +, x, ys...)`, without intermediate allocations.
+
+Does not accept a `dims` keyword.
+"""
+logprod(f, x) = logprod(Iterators.map(f, x))
+logprod(f, x, ys...) = logprod(f(xy...) for xy in zip(x, ys...))
+
+# Iterator version, uses the same `_logprod_op`, should be the same speed.
+function logprod(x)
+    iter = iterate(x)
+    if isnothing(iter)
+        T = Base._return_type(first, Tuple{typeof(x)})
+        return T <: Number ? zero(float(T)) : 0.0
```

Comment:
No internals, please. `prod` works for empty iterators and arrays, so we could just use
```suggestion
        y = prod(x)
        return log(abs(y)), sign(y)
```

Reply:
Agreed, this was left over - I didn't mess with the iterator version much yet